models/issue_comment.go

```diff
@@ -566,6 +598,141 @@ func CreateRefComment(doer *User, repo *Repository, issue *Issue, content, commi
 	return err
 }
 
+// ClearPullPushComment clear all the push commit on issue since fore push
+func ClearPullPushComment(issue *Issue) error {
+	_, err := x.Delete(&Comment{
+		Type:    CommentTypePullPushCommit,
+		IssueID: issue.ID,
+	})
+	return err
+}
+
+func createPullPushCommentsNonForcePush(doer *User, issues []*Issue, commits []*PushCommit) error {
+	sess := x.NewSession()
+	defer sess.Close()
+
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	for _, issue := range issues {
+		for i := len(commits) - 1; i >= 0; i-- {
+			_, err := createComment(sess, &CreateCommentOptions{
+				Type:      CommentTypePullPushCommit,
+				Doer:      doer,
+				Repo:      issue.Repo,
+				Issue:     issue,
+				CommitSHA: commits[i].Sha1,
+				Content:   commits[i].Message,
+			})
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return sess.Commit()
+}
+
+// CreatePullPushComments creates commit comments when push commits to a pull request.
+func CreatePullPushComments(doer *User, issues []*Issue, commits []*PushCommit, isForcePush bool) error {
+	if !isForcePush {
+		return createPullPushCommentsNonForcePush(doer, issues, commits)
+	}
+
+	sess := x.NewSession()
+	defer sess.Close()
+
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	var commitIDCache = make(map[string]string)
+	var ok bool
+	for _, issue := range issues {
+		var commitID string
+		var key = fmt.Sprintf("%s/%s", issue.Repo.RepoPath(), issue.PullRequest.BaseBranch)
+		if commitID, ok = commitIDCache[key]; !ok {
+			gitRepo, err := git.OpenRepository(issue.Repo.RepoPath())
+			if err != nil {
+				return fmt.Errorf("OpenRepository: %v", err)
+			}
+			commit, err := gitRepo.GetBranchCommit(issue.PullRequest.BaseBranch)
+			if err != nil {
+				return fmt.Errorf("GetBranchCommit: %v", err)
+			}
+			commitID = commit.ID.String()
+			commitIDCache[key] = commitID
+		}
+
+		var startIdx = -1
+		for i := len(commits) - 1; i >= 0; i-- {
+			if commitID == commits[i].Sha1 {
+				startIdx = i
+				break
+			}
+		}
+		if startIdx > -1 {
+			commits = commits[:startIdx]
```

Comment:
From what I can tell, this is not right.

Suppose I have the following (where later-in-the-alphabet commits have a later timestamp; it appears this affects the order of `opts.Commits`, which is itself kind of concerning)
```
A---B(master, origin/master)
 \--C(develop, origin/develop)
```
and I rebase develop onto master, resulting in 
```
      --C'(develop)
     /
A---B(master, origin/master)
 \--C(origin/develop)
```

If I force push develop, the order of `opts.Commits` will be `[C', C, B]`. In this case (since `BaseBranch` is master), the start index will be 2, and both `C'` and `C` will be included in the `commits` variable.

Reply:
Also, it is possible that the head commit of the base branch is in `commits` at all.

Suppose I have
```
A(master, origin/master)
 \
  B(develop, origin/develop)
```

and I replace commit `B` with a new commit `C` locally
```
  C(develop)
 /
A(master, origin/master)
 \
  B(origin/develop)
```

If I force push develop, `opts.Commits` will contain `[C, B]`, in that order. In this case, `startIdx` will still be -1 after the for loop, and `commits` will still contain both `C` and `B`, which is incorrect.

Reply:
C will not in the commits list since it has been commit before. See https://github.com/go-gitea/gitea/blob/master/models/update.go#L257

Reply:
@lunny I have added print statements to the beginning of `CommitRepoActions` (which is run after line 257 of `models/update.go`), to print the contents in `opts.Commits`. In both of the cases described in my comments, I have gotten the results that match what my comments describe. If you are unable to reproduce these results, or if I am missing something, please let me know.